PSKoans/Koans/Constructs and Patterns/AboutRegularExpressions.Koans.ps1

```diff
@@ -0,0 +1,589 @@
+        }
+
+        It 'does not find just one badger' {    
+            $badgers = 'BadgerBadgerBadger'
+            $groupingPattern = '(____){__}'
+            [regex]::Matches($badgers, $groupingPattern).Value | Should -Be 'BadgerBadgerBadger'
+            $groupingPattern.Contains(')') | Should -BeTrue    # Use a group!
+        }
+    }
+
+    Context '[ and ] - Square Brackets' {
+
+        It 'defines a set' {
+            <#
+                Square brackets denote a set or collection of symbols within a regular
+                expression. Imagine a pattern that might read "a or b or c or d". You may use square
+                brackets to create that regex. The set contained within the square brackets
+                represent one charactermwithin the string being searched.
+
+                The fancy regex name for this is a "character class".
+            #>
+    
+            $____ | Should -Be 'End of the line' -match '[efg]$'
+        }
+
+        It 'wants you to end the string a certain way' {    
+            '____' -match '[abcdefg]$' | Should -BeTrue
+        }
+
+        It 'wants you to avoid ending the string a certain way' {
+            <#
+                Normally, the ^ character means "start of a line", but inside of square brackets, it
+                negates the set. Instead of "these symbols" it means "not these symbols"
+            #>
+            '____' -match '[^abcdefg]$' | Should -BeTrue
+        }
+
+        It 'still needs to match a ^ inside a set' {    
+            <#
+                What do you think you might do if you want to match a literal ^ symbol within a set?
+                Use the ^ (caret) symbol in this one.
+            #>
+            $caretMatch = '[____]'
+            '^&*' -match $caretMatch | Should -BeTrue
+            $caretMatch.Contains('^') | Should -BeTrue # Use the caret symbol!
+        }
+    }
+}
+
+Describe 'Meditative Examples' {
+    # Here are some challenges for you to put your new skills to work on.
+    
+    Context 'Isolate a Username From a domain\username String Using Just One Regex Pattern' {
+
+        It 'is handy for isolating parts of strings' {
+            $usernamePattern = '____'
+            [regex]::Matches('cat\harley', $usernamePattern).Value | Should -Be 'harley'
+            [regex]::Matches('cat\kali', $usernamePattern).Value | Should -Be 'kali'
+            [regex]::Matches('human\thomas', $usernamePattern).Value | Should -Be 'thomas'
+        }
+    }
+
+    Context 'Validate a Bunch of Phone Numbers' {
+        
+        <#
+            Validate a bunch of phone numbers - get rid of the non-numeric characters and check
+            which ones are actually the right length to be dialed.
+        #>
+
+        $phoneNumbers = @(
+            '1 425 555 1234',
+            '1-425-555-4321',
+            '1.425.555.6789',
+            '01234'
+            '+14255556789'`
+        )
```

Comment:
Unnecessary commas and a backtick, for some reason. 😋

Reply:
100% that backtick is from my Ergodox layout. Little accidental backticks everywhere. Just usually they're so disruptive that I notice them 😂